### PR TITLE
add fb algorithm, pnp-prior functionality, abstract classes, and kera…

### LINKED ISSUE #5: Implement simple forward-backward solver

### LINKED ISSUE #6: Investigate any refactoring or architecture improvements

### PR DESCRIPTION
**Pull request by CosmoMatt**
Overview
This PR includes quite a few bits and pieces that have been outstanding. Primarily it adds forward backward solver and enhances the codebase by supporting keras/tensorflow likelihood/priors.

### Issues:
closes #5 
closes #6 

### Additions:
- adds a basic forward backward solver (this has been tested externally and can function with keras priors/gradients)
- adds abstract classes to linear operators, proximal operators, ai operators, gradient operators.
- adds forward backward 1d notebook 
- updates custom operator notebook to inherit the corresponding abstract class

### CODE REVIEW

File: optimusprimal/primal_dual.py [hunk 1]
```diff
@@ -6,18 +6,21 @@
...
+):
     """Evaluates the Primal dual forward backward optimization
 
     Args:
 
         x_init (np.ndarray): First estimate solution
         options (dict): Python dictionary of optimisation configuration parameters
         g (Grad Class): Unconstrained data-fidelity class
-        f (Prox Class): Constrained data-fidelity class
+        p (Prox Class): Constrained data-fidelity class
```
Comment (MatthijsMars) [inline optimusprimal/primal_dual.py:19]: should keep order the same as arguments in function call

**Review by MatthijsMars — APPROVED** [review]

**Comment by tobias-liaudat** [comment]
Hi @CosmoMatt and @MatthijsMars 
I was checking out this PR, is there a reason it has not been merged?
